RFC: Implementable trait aliases

Jules-Bertholet opened the pull request:
Allow writing `impl` blocks for certain trait aliases.

[Prior discussion on Internals](https://internals.rust-lang.org/t/implementable-trait-aliases/18806/)

@rustbot label A-traits

programmerjake commented on text/3437-implementable-trait-alias.md:
```diff
@@ -0,0 +1,319 @@
+- Better ergonomics compared to purely proc-macro based solutions.
+- One alternative is to allow marker traits or auto traits to appear in `+` bounds of implementable aliases.
+(For example, `trait Foo = Bar + Send;` could be made implementable). However, I suspect that the complexity would not be worthwhile.
+- Another possibility is to require an attribute on implmenentable aliase; e.g. `#[implementable] trait Foo = ...`. Again, I don't think that the complexity is warranted.
```
one benefit of requiring `#[implementable]` is that it mitigates the confusing aspect of:
```rust
#[implementable] // error: must only have one trait on rhs of equal -- generated because of #[implementable]
pub trait Foo = Bar + Baz;
// vs.
#[implementable]
pub trait Foo = Bar
where
    Self: Baz;
```

Jules-Bertholet replied:
I've added some more discussion about this.

burdges commented:
As an aside, it'd be nice if type aliases permitted instantiation:
```
pub struct FooT<T>([T; 32]);
pub type Bar = Foo<u8>;

impl Default for Bar {
    fn default() -> Bar {
        Bar( Default::default() )   // Forbidden right now.
    }
} 
```

programmerjake commented:
> As an aside, it'd be nice if type aliases permitted instantiation:

they do, if you use curly braces (except for tuples):
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b95fe39281037ac567cfc8dfa7aba44b

i think it'd also be nice to allow curly brace syntax for tuples for ease of writing proc macros so there's one consistent syntax that works on all struct-like types:
```rust
type T<A, B> = (A, B);
let v = T::<_, _> { 0: "abc", 1: 123 };
match v {
    T::<_, _> {0: a, 1: b} => println!("it works! a={a} b={b}"),
}
```

dlight commented:
> they do, if you use curly braces (except for tuples):
> https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b95fe39281037ac567cfc8dfa7aba44b

Wait, so `T { 0: 5 }` works but `T(5)` doesn't? This sounds like a bug.

Jules-Bertholet commented:
> Wait, so `T { 0: 5 }` works but `T(5)` doesn't? This sounds like a bug.

Changing it behavior would be a breaking change, for code like the following:

```rust
struct Foo();
type Bar = Foo;
fn Bar() {}
```

Anyway, this is all a bit off-topic for the RFC :grin:

nielsle commented:
I think that the RFC should mention #1672 and mutually exclusive traits

https://github.com/rust-lang/rust/issues/20400
https://geo-ant.github.io/blog/2021/mutually-exclusive-traits-rust/
https://stackoverflow.com/questions/57749827/mutually-exclusive-traits

Is this RFC an alternative to #1672 or is it orthogonal?

Jules-Bertholet commented:
@nielsle I don't see any relationship between this RFC and that issue.

tmandry commented:
I think the summary motivation section should be fleshed out more, ideally with real-world examples (possibly simplified). After looking through the text this _does_ seem useful but it wasn't obvious at all to me what the feature _was_ from those sections.

Jules-Bertholet commented:
I've specified that implementable trait aliases also support fully-qualified method call syntax.

Veykril commented on text/3437-implementable-trait-alias.md:
```diff
@@ -287,6 +287,25 @@ impl Frobber for MyType {
 
 Trait aliases are `unsafe` to implement iff the underlying trait is marked `unsafe`.
 
+Implementable trait aliases can also be used with trait-qualified and fully-qualified method call syntax. When used this way,
```
Why only function calls? This should probably state that they can be used in general as the qualifying trait for a qualified path, so that you address consts, functions and type aliases altogether.

Jules-Bertholet replied:
Yes, that would make sense, I will fix it.

nikomatsakis commented:
@rustbot +I-lang-nominates +I-types-nominated

I'm nominating this for @rust-lang/lang and @rust-lang/types discussion. I myself am in favor of this RFC. I've seen a lot of demand for a simplified version of trait aliases where--

(A) We only support `trait Foo = Bar + Baz` format, essentially.
(B) The trait `Foo` is implementable and, in general, acts like a "normal trait" so that users don't have to know about the more detailed version.

I expect this to be relevant to async fn in traits as well (cc @tmandry) it's quite common to have something like `trait Service` (which requires `Send`) and `trait LocalService` (which does not).

Restricting to point (A) has the advantage of avoiding some complex corner cases. Note though that I do want to support the "inline bound" syntax, so that you could do `trait SendIterator = Iterator<Item: Send>` (as well as RTN if we adopt that), so that trait aliases can put bounds on associated types.

nikomatsakis commented:
I guess that my question to the lang/types teams, respectively, are:

1. Lang: what do people think about the motivation here as well as the more limited proposal? Are there important trait alias use cases not covered there?
2. Types: are there concerns about the limited proposal that I put forward?

nikomatsakis commented:
One though -- @Jules-Bertholet -- I'm not sure if the RFC covers it, I didn't have time to read in super detail, but I think that if you are implementing `trait Foo = Bar + Baz`, that implies "unioning" the methods from `Bar` and `Baz` into one impl block. Does the RFC discuss this?

I have, for a long time, wanted the ability to implement a trait **and** its supertraits together in a single impl block, I wonder if it's worth thinking about that as well, though I'd probably want to separate it out from this RFC.

Jules-Bertholet commented:
> One though -- @Jules-Bertholet -- I'm not sure if the RFC covers it, I didn't have time to read in super detail, but I think that if you are implementing `trait Foo = Bar + Baz`, that implies "unioning" the methods from `Bar` and `Baz` into one impl block. Does the RFC discuss this?

The RFC does not do any sort of trait unioning like this, I hadn't even considered it as a possibility (I will add it to the alternatives section). Notably, you would need to handle name collisions.

Jules-Bertholet commented:
I've added text to the alternatives section addressing @nikomatsakis's "unioning" idea.

nikomatsakis commented:
@Jules-Bertholet 

>  Notably, you would need to handle name collisions.

Yes. Presumably that would be an error.

Jules-Bertholet commented:
> Presumably that would be an error. 

That would have backward-compatibility implications, if one of the traits adds a defaulted method that conflicts. Technically "minor" breakage, but still not ideal.

nikomatsakis commented:
@Jules-Bertholet good callout -- I think the error would have to be if your impl contains a single function item that is ambiguous.

scottmcm commented:
First impression: I'd love to have a way to split an existing trait into two (to pull out a more limited version) that wasn't a breaking change.  So big 👍 from me on doing *something* that can help that kind of thing.

I'm also a fan of making trait aliases *feel* more like traits, since right now I tend to think of them as "bounds aliases" since you can't use them in a bunch of places where you can use a "normal" trait.  (Unlike type aliases, which can be used in most (all?) places where normal types can be used.)

I have a couple of worries after skimming through this, though.

My biggest worry is in interactions with the accepted <https://rust-lang.github.io/rfcs/2845-supertrait-item-shadowing.html>.  That made adding a method to a trait less of a breaking change again, but I think that this would effectively undo it -- anyone implementing through a trait alias like this would be broken in those cases, I think.  The "obvious" fix there would be to have people write which trait the item they're implementing came from, but that violates my wish from the first paragraph in this comment about being able to split traits, so I don't like it at all.  As such I'm not sure how best to address this.  Maybe some kind of coherence restriction on what traits can be combined could work?  Not sure.

I'm also worried about the "can you paste it into an impl block" rule of thumb.  As the RFC mentions about syntactic distance, having "the same" trait alias behave differently depending whether it's written with `+` or `where` feels quite surprising to me.  TBH, I feel like if I wrote `trait Qux = Foo + Bar;` and I'd heard that trait aliases were implementable, I'd expect to be able to `impl Qux for MyType { ... }` and have it just work to write the methods in that block.  Having it unclear whether the author *intended* for it to be implementable feels much like `dyn` to me, where I often wish there was a way to be more intentionable about what you expect consumers to do with the trait.

Spitballing:
- Should maybe `trait Foo = ` *always* be implementable?  And use a different construct for bounds aliases that aren't implementable in the same way traits are?
- What constraints exist to allowing something like `impl Foo+Bar for MyType` or `impl Ord for MyType { fn cmp(...; fn partial_cmp(...; }`?  Could we find rules that allow this for more cases?  Is there some horrible downside to it we could put in a design note about why we'll never do that?

Jules-Bertholet commented:
> My biggest worry is in interactions with the accepted <https://rust-lang.github.io/rfcs/2845-supertrait-item-shadowing.html>.  That made adding a method to a trait less of a breaking change again, but I think that this would effectively undo it -- anyone implementing through a trait alias like this would be broken in those cases, I think.

Is this concern meant to apply only to the extended proposal that allows multiple "primary" traits? The RFC in its current state would not have this issue.

nikomatsakis commented:
We discussed this some in our @rust-lang/lang meeting today. We didn't come to a full consensus but we raised a few points:

* People in meeting seemed generally positive about being able to implement a subtrait and supertrait at the same time.
* The [where the where GAT discussion](https://rust-lang.github.io/generic-associated-types-initiative/design-discussions/where-the-where-1.html#what-about-trait-aliases-dont-they-distinguish-where-clause-placement), in my personal view, effectively deprecated the `where Foo = where Bar` syntax, so I don't think we should lean on that.
* We were also interested in saying "you can't implement every trait alias, but you can do it a lot of the time". 

To the last point, I was floating specifically the idea that we can use the same rules that we use to define when one method takes precedence over another to decide what method is being *implemented*.

For example, if you had

```rust
trait Foo { fn method1() { ... }}
trat Bar: Foo { fn method2() { ... }}

impl Bar { 
    fn method1() { ... } // from Foo
    fn method2() { ... } // from Bar
}
```

and then `Foo` adds a `fn method2() { ... }`, we could say that the impl is still legal but the `Foo::method2` is defaulted.

(Of course, if `Bar` added a `fn method1() { ... }`, that would be a breaking change, but this is true anyway, because code that invokes `x.method1()` would now be calling `Bar::method1` and not `Foo::method1`. [Fragile base classes](https://en.wikipedia.org/wiki/Fragile_base_class) are hard.)

In general, my preference would be to say that `trait Foo: Bar + Baz` and `trait Foo = Bar + Baz` are basically the same and usable in all the same places in all the same ways.

Jules-Bertholet commented:
> People in meeting seemed generally positive about being able to implement a subtrait and supertrait at the same time.

This may or may not be a good idea, but in either case I think it is out of scope for this RFC.

> The [where the where GAT discussion](https://rust-lang.github.io/generic-associated-types-initiative/design-discussions/where-the-where-1.html#what-about-trait-aliases-dont-they-distinguish-where-clause-placement), in my personal view, effectively deprecated the `where Foo = where Bar` syntax

What syntax, exactly? I've improved the language around how `where` clauses are treated by this RFC, hopefully that addresses your point?

---

I've added another restriction on which trait aliases are implementable, to avoid problems around overlapping impls. Generic parameters of implementable type aliases now need to be used as generic parameters to the alias's primary trait.

scottmcm commented:
Another random thought: it would be nice if `trait SelfAdd = Add<Output = Self>;` was implementable as well.  That would give a path to moving a type to an associated one without breaking things, and just be convenient for cases where there's a particularly-common choice for one or more associated types.

Jules-Bertholet commented:
> Another random thought: it would be nice if `trait SelfAdd = Add<Output = Self>;` was implementable as well.

It is intended to be. The RFC has an example like this already, but I will improve the language to make it more clear that this is allowed.

Jules-Bertholet commented:
I've made some changes in response to the design meeting ([Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/410673-t-lang.2Fmeetings/topic/Design.20meeting.202023-10-25), [HackMD](https://hackmd.io/VGOOy3mnQ6SR85ex0NsiCg?view)) held earlier today. Notably, I've:

- Clarified the language around how and where `where` bounds are enforced.
   - The new rules bring this RFC more or less in line with @nikomatsakis's model of "desugaring to a blanket impl."
- Removed the restriction that implementable trait aliases must use all generic parameters as parameters of their primary trait.
- Added a few more sentences on why the syntax is the way it is.

One thing I haven't yet updated is the motivation section. Many people have pointed out that they would like motivating examples that don't rely on unstable features. I plan to get to that Soon™—but in the meantime, if anyone has a good real-life example, feel free to share.

Also in the design meeting, many people had thoughts on this RFCs relationship to `LendingIterator`/GATification. Because that part of the RFC is so speculative, I don't think it would be wise to expand on it here, but perhaps we should discuss this in another forum (Zulip or Internals?).

nikomatsakis commented:
I found the motivation in the RFC a bit off-target. I think that the 'weak-strong' concept is correct, but not really *specific* to trait aliases. Every subtrait corresponds to a "refined" view of the supertrait (sometimes with extra capabilities/methods as a result of that refinement). I think of trait aliases as really just a *shorthand* for declaring refined traits in cases where no "per-type behavior" is needed (i.e., because there are no new methods, or even (eventually) because the definition of the methods is always the same). 

To my mind, the real motivation here is expanding on the idea of trait aliases as being able to provide a "complete abstraction", where users don't have to know about the "contents" of the alias, they can treat it "as if" it were a standalone trait. I think the RFC started in the right vein there -- trait aliases are used to provide a "simplified, convenient view" onto a more complex underlying reality, but they're not able to truly hide that from users.

Eventually I would like it if users could do the same for *any* trait -- i.e., if I have a trait `Foo: Bar`, I should be able to implement `Foo` and `Bar` together etc. But I don't think that has to be in this RFC, and it'll take a bit of thought to figure out how it should work best. Still, the intution makes sense to me -- i.e., if I am in some OOP language, I don't have to separately define Shape and Rectangle and Square, I can just define Square and that Rectangle/Shape come "for free" with that.

Jules-Bertholet commented:
> I think of trait aliases as really just a shorthand for declaring refined traits in cases where no "per-type behavior" is needed (i.e., because there are no new methods, or even (eventually) because the definition of the methods is always the same.

I agree with this framing, that's a good way of putting it.

Jules-Bertholet commented:
Having given it some more thought, I now suspect that associated types in trait aliases could be sufficient to GATify `Iterator` *without* a separate "variance bounds" feature:

```rust
pub trait LendingIterator {
    type LentItem<'a>
    where
        Self: 'a;

    fn next<'a>(&'a mut self) -> Option<Self::LentItem<'a>>;
}

pub trait Iterator = LendingIterator
where
    // Still need to resolve implied `'static` bound problem
    // (https://blog.rust-lang.org/2022/10/28/gats-stabilization.html#implied-static-requirement-from-higher-ranked-trait-bounds)
    for<'a> Self::LentItem<'a> = Self::Item,
{
    type Item;
}
```

Jules-Bertholet commented:
>  associated types in trait aliases could be sufficient to GATify Iterator without a separate "variance bounds" feature

I've added a [section](https://github.com/Jules-Bertholet/rfcs/blob/implementable-trait-alias/text/3437-implementable-trait-alias.md#trait-aliases-constrained-by-their-associated-items) to the future possibilities discussing this.

Jules-Bertholet commented:
I've rewritten the AFIT example in the motivation section in terms of the [`trait-variant`](https://crates.io/crates/trait-variant) crate.

Jules-Bertholet commented:
There's one issue I hadn't noticed before that limits the applicability of this feature to a degree. Currently, if you have:

```rust
pub trait Foo { /* ... */ }
pub trait SendFoo: Foo + Send {}
impl<T: Foo + Send + ?Sized> SendFoo for Foo {}
```

Then,  `dyn Foo + Send` and `dyn SendFoo` are distinct types. Migrating to:

```rust
pub trait Foo { /* ... */ }
pub trait SendFoo = Foo where Self: Send;
```

Makes these into the same type, which is a breaking change (trait impls could start overlapping). This isn't a problem for the async `Send` variant use-case, as traits with async methods aren't (yet) object safe, but it could potentially make migrating from blanket impls to implementable trait aliases a breaking change in some other cases.

willmtemple commented:
@Jules-Bertholet I have another use case that I haven't seen mentioned here: I want to be able to hide a trait behind a module using a proc macro. I want to do this to create a sort of RTTI system for tracking which types implement traits dynamically (along with constructing other forms of metadata about the implementation). This is all in service of building a scripting language on top of Rust.

Here's a really simplified example:

```rs
#[runtime_trait]
pub trait Foo {
  fn bar(&self) -> i32;
}
```

would be rewritten to

```rs
#[doc(hidden)]
trait __Foo {
  fn bar(&self) -> i32;
}

pub mod Foo {
  pub trait Foo = super::__Foo;
  
  #[::linkme::distributed_slice]
  pub static IMPLEMENTORS: [TypeId];
}
```

And the impl:

```rs
#[runtime_trait_impl]
impl Foo for T {
  fn bar(&self) -> i32 { 3 }
}
```

would be rewritten to:

```rs
// Relies on being able to impl __Foo through the trait alias inside the module.
impl Foo::Foo for T {
  fn bar(&self) -> i32 { 3 }
}

#[::linkme::distributed_slice(Foo::IMPLEMENTORS)]
static __FOO_T: TypeId = TypeId::of<T>();
```

- The trait itself cannot be declared inside the `Foo` module, since the module does not inherit scope from the outer module, and that will break the written definition of the trait's constraints and items.
- If I give the module a different name from the trait, the user has to import an "invisible" module generated by the proc macro anywhere they want to use `#[protocol_impl]`, and that would suck.
- I've removed a lot of complexity around handling generic traits, but the general idea could hold for generic traits as well.

Kixunil commented:
I have a real-world example as well. My [`push_decode`](https://docs.rs/push_decode) crate provides an abstraction over both sync/async and std/no_std encoding. This is achieved simply by requiring the implementors to provide an encoder - a state machine that implements the encoding. I'd love to also provide traits that take care of encodable objects. However there are various strategies used to encode things. e.g. `u32` can be encoded as littel endian, big endian, varint... So what I want is this:

```trait
// Analogous to serde::Serialize
pub trait Encode<Strategy> {
    type Encoder: Encoder;
    // additional things like const MIN_SERIALIZED_LEN: usize
    fn encode(&self) -> Self::Encoder;
    // various helper methods taking advantage of the additional items. E.g. fn encode_to_vec(&self) which reserves MIN_SERIALIZED_LEN
}
```

Then each protocol can define `enum ProtocolName {}` and impl `Encode<ProtocolName>`. However crates with many encodable types will have to write `Encode<ProtocolName>` everywhere and import `ProtocolName`. This is annoying and it'd be much nicer to define an alias `trait Encode = push_bytes::Encode<ProtocolName>` and then just use `Encode` everywhere.

tmandry commented on text/3437-implementable-trait-alias.md:
```diff
@@ -0,0 +1,743 @@
+  - However, trait transformers would also address more use-cases (for example,
+    sync and async versions of a trait).
+- Better ergonomics compared to purely proc-macro based solutions.
+- One alternative is to allow marker traits or auto traits to appear in `+`
+  bounds of implementable aliases. (For example, `trait Foo = Bar + Send;` could
+  be made implementable).
```
I'm in favor of this, because while the symmetry with impl blocks is nice I think the symmetry with bounds is ultimately more important (and more confusing to break). If other lang team members agree I think we should go ahead and support it on day 1.

Otherwise, given that this is something we can allow in the future, can it be listed as a future possibility?

nikomatsakis commented:
We [discussed this RFC in our triage meeting today](https://hackmd.io/TpvcFSJoRhGdydaakNXcZg?view#RFC-Implementable-trait-aliases-rfcs3437). 

The meeting consensus was that we'd ideally like to have an RFC that lays out a forward looking plan that goes a bit further than the specifics in this RFC:

* It should be possible to implement a trait and its supertraits together
* `trait Alias<T..> = Bound.. where WC...` should be equivalent to the pair of
    * `trait Alias<T..>: Bound.. where WC... {}`
    * `impl<Self: Bound.., T..> Alias<T..> for Self where WC...` 
    * If there are no where clauses, this is true today:
        * `forall T. T: Alias => T: Bound..` because of the supertraits
        * `forall T. T: Bound.. => T: Alias` because of the impl
    * I don't think *THIS RFC* should specify the behavior for WC, only that we want the equivalence to hold. The correct behavior of where-clauses for trait definitions (in particular the implied bounds expansion) can be covered separately.
* Combining these two, `impl Alias for Type` should work and be equivalent to `impl (Bound..) for Type

Together (I think) these changes also allow for other transformations. It'd be good to capture these as part of the RFC's motivation. In addition to the "trait alias" = "trait + impl" equivalence above, the other equivalence we had in mind was being able to convert `trait Foo { fn a(); fn b(); }` into `trait Foo: Bar { fn a(); } trait Bar { fn b(); }` without breaking existing impls.

Note that just because the RFC includes this content doesn't mean we have to stabilize or even implement all of it at once. I think it'd be useful to identify "tiers" of support. In particular

*Trait aliases that include bounds with one primary trait + auto or marker traits but no where-clauses*

seems like an important tier that more-or-less works today.

nikomatsakis commented:
This wasn't discussed in the meeting, but I personally would go further and say that my ideal is that you can do `impl $Bound for Ty` for any bound (i.e., anything that could come after a `:`). That would mean `impl Debug + Display for Type` etc. Basically, if you can do `trait Subtrait: Debug + Display` and then `impl Subtrait for Type` and have that cover all 3, why not be able to say it explicitly?

Jules-Bertholet commented:
It seems that the lang team wants this RFC to be more broad, so I have gone ahead and done that. You can now implement trait aliases that combine multiple traits. In addition, you can specify bodies for trait aliases, that define additional names for different trait items.

Jules-Bertholet commented:
> It's worth mentioning that we're also talking about allowing impls that cover both a subtrait and supertrait, as described here:
> 
> * https://github.com/rust-lang/rust/pull/135881#issuecomment-2718417230

I don't think I agree with doing it in exactly that way, but that is definitely relevant, and I will update the RFC with that example and how my proposal would handle it.

programmerjake commented on text/3437-implementable-trait-alias.md:
```diff
@@ -0,0 +1,859 @@
+    // so `Item` must be `i32`
+    // and we don't need to specify it.
+
+    fn next(&mut self) -> i32 {
+        -27
```
```suggestion
    fn next(&mut self) -> Option<i32> {
        Some(-27)
```

programmerjake commented on text/3437-implementable-trait-alias.md:
```diff
@@ -0,0 +1,859 @@
+    // The alias constrains `Self::Item` to `i32`, so we don't need to specify it
+    // (though we are allowed to do so if desired).
+
+    fn next(&mut self) -> i32 {
+        -27
```
```suggestion
    fn next(&mut self) -> Option<i32> {
        Some(-27)
```

programmerjake commented on text/3437-implementable-trait-alias.md:
```diff
@@ -0,0 +1,859 @@
+let iter = [1_u32].into_iter();
+let _: IntIter::Item = IntIter::next(&mut iter); // works
+let _: <array::IntoIter as IntIter>::Item = <array::IntoIter as IntIter>::next(); // works
+//IntIter::clone(&iter); // ERROR: trait `Iterator` has no method named `clone()`
```
right below you state items from traits in `where` clauses are accessible, so `IntIter::clone` should be accessible

programmerjake commented:
> It's worth mentioning that we're also talking about allowing impls that cover both a subtrait and supertrait, as described here:
> 
> * [Arbitrary self types v2: stabilize rust#135881 (comment)](https://github.com/rust-lang/rust/pull/135881#issuecomment-2718417230)

I think that could be achieved with a trait alias (assuming `dyn TraitAlias` works when the trait alias is `NonAutoTrait1 + NonAutoTrait2` -- because the alias is a named item, it can be treated as a normal trait when rustc is figuring out what vtable to generate):

```rust
pub trait Receiver {
    type Target: ?Sized;
}

// what Deref resolves to in edition >= 2027
pub trait e2027#Deref: Receiver {
    fn deref(&self) -> &Self::Target;
}

// what Deref resolves to in edition <= 2024
pub trait e2024#Deref = Receiver + e2027#Deref {
    type Target: ?Sized = <Self as Receiver>::Target;
    fn deref = <Self as actual_deref>::deref;
}
```

Jules-Bertholet commented:
> (assuming `dyn TraitAlias` works when the trait alias is `NonAutoTrait1 + NonAutoTrait2` -- because the alias is a named item, it can be treated as a normal trait when rustc is figuring out what vtable to generate)

I’ve added a section on this. TLDR: no, it does not work in general, but if one of the non-auto traits is a subtrait of all the others then it does (`dyn Alias` is then just a synonym of `dyn Subtrait`).

programmerjake commented on text/3437-implementable-trait-alias.md:
```diff
@@ -0,0 +1,959 @@
+}
+
+trait Alias = Foo {
+    type Associated = Self::Assoc;
```
are you required to write `type Associated: ?Sized = Self::Assoc` if you want to avoid a default `Sized` bound when defining a trait alias?

Jules-Bertholet replied:
No, there is not, just as there is no such default `Sized` bound for type aliases in other contexts.

programmerjake replied:
there *are* default bounds for associated types, which can be written with basically the same syntax, so I think this definitely needs to be specified.
```rust
trait Tr<T: Deref> {
    type Ty = <T as Deref>::Target; // has a Sized bound
}
trait Tr2<T: Deref> = Send {
    type Ty = <T as Deref>::Target; // has no Sized bound
}
```

Jules-Bertholet replied:
I’ve added a note in the RFC.

Jules-Bertholet commented:
I’ve added a small note to the future possibilities, connecting this to const traits. I will try to discuss target features as well, but need some time to think about that first.

Jules-Bertholet commented:
I’ve now realized that this feature is actually sufficient to address the “make `LendingIterator` a supertrait of `Iterator`” issue on its own. I’ve moved that out of the future possibilities and into the main body of the RFC.

Jules-Bertholet commented:
> I will try to discuss target features as well, but need some time to think about that first.

This has now been added to the future possibilities.

Jules-Bertholet commented:
I’ve removed the implementability restrictions on `type` and `const` alias items. They just added pointless complexity.

Jules-Bertholet commented:
> It's worth mentioning that we're also talking about allowing impls that cover both a subtrait and supertrait, as described here:
> 
> * https://github.com/rust-lang/rust/pull/135881#issuecomment-2718417230

I’ve added a section on why I believe making this “just work” in general would be a bad idea, and why the extra hoops that this RFC requires are necessary.

Jules-Bertholet commented on text/3437-implementable-trait-alias.md:
```diff
@@ -0,0 +1,1088 @@
+The library author may want to rename the trait and its items to something less
+unwieldy. Unfortunately, he has no good way to accomplish this at present.
+
+# Guide-level explanation
```
Unfortunately, it would not, because of the default impl of `clone_from` in terms of `clone`, which the mechanisms in this RFC wouldn’t be able to preserve. However, I have a [post on Internals](https://internals.rust-lang.org/t/idea-partial-impls/22706) from 2 months ago, that would offer a solution. Combining this RFC and that post’s features, you could do:


```rust
trait CloneInto: CloneFrom<Self> {
    fn clone(&self) -> Self;
}
trait CloneFrom<T: ?Sized> {
    fn clone_from(&mut self, other: &T);
}

// Leverages this RFC
trait Clone = CloneInto + CloneFrom;

// Leverages the Internals post
default partial impl<T: CloneInto> CloneFrom<Self> for T {
     fn clone_from(&mut self, source: &Self) { *self = source.clone(); }
}
```

Jules-Bertholet commented on text/3437-implementable-trait-alias.md:
````diff
@@ -0,0 +1,1091 @@
+}
+```
+
+This could be used to add `LendingIterator` as a supertrait of `Iterator`, as
+mentioned in the motivation section:
+
+```rust
+pub trait LendingIterator {
+    type Item<'a> where Self: 'a;
+    fn next(&mut self) -> Option<Self::Item<'_>>;
+}
+
+pub trait Iterator = for<'a> LendingIterator<Item<'a> = <Self as Iterator>::Item> {
+    type Item;
+}
+```
+
+You’ll note that `Iterator`’s body does not explicitly define `Item`. Instead,
+it’s defined implicitly by the `for<'a> <Self as LendingIterator>::Item<'a> =
+<Self as Iterator>::Item` clause in the alias definition.
````
I don’t have a solution for that issue, that’s not part of this RFC. The unwanted `'static` bound is a serious impediment to using `LendingIterator` even without bringing `Iterator` into the picture. A solution to the problem would have to be found before bringing the trait into the standard library anyway; whatever that solution is, could be leveraged by the alias as well.

Jules-Bertholet replied:
It *could* be right. It depends on whether you see the `'static` bound as a feature or a bug

Jules-Bertholet replied:
I added a note to the RFC.
